src/upgrades/mod.rs

```diff
@@ -86,30 +146,101 @@ impl From<crate::commons::error::Error> for UpgradeError {
 
 impl std::error::Error for UpgradeError {}
 
+//------------ DataUpgradeInfo -----------------------------------------------
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct DataUpgradeInfo {
+    pub last_event: u64,
+    pub last_command: u64,
+    pub last_update: Time,
+}
+
+impl Default for DataUpgradeInfo {
+    fn default() -> Self {
+        Self {
+            last_event: 0,
+            last_command: 0,
+            last_update: Time::now(), // will be overwritten to appropriate value
+        }
+    }
+}
+
+impl From<&DataUpgradeInfo> for StoredValueInfo {
+    fn from(upgrade_info: &DataUpgradeInfo) -> Self {
+        StoredValueInfo {
+            snapshot_version: 0,
+            last_event: upgrade_info.last_event,
+            last_command: upgrade_info.last_command,
+            last_update: upgrade_info.last_update,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum UpgradeMode {
+    PrepareOnly,
+    PrepareThenFinalise,
+}
+
 //------------ UpgradeStore --------------------------------------------------
 
 /// Implement this for automatic upgrades to key stores
 pub trait UpgradeStore {
     fn needs_migrate(&self) -> Result<bool, UpgradeError>;
-    fn migrate(&self) -> Result<(), UpgradeError>;
 
-    fn version_before(kv: &KeyValueStore, later: KrillVersion) -> Result<bool, UpgradeError> {
-        kv.version_is_before(later).map_err(UpgradeError::KeyStoreError)
+    fn prepare_new_data(&self, mode: UpgradeMode) -> Result<(), UpgradeError>;
+
+    fn version_before(&self, later: KrillVersion) -> Result<bool, UpgradeError> {
+        self.store()
+            .version_is_before(later)
+            .map_err(UpgradeError::KeyStoreError)
     }
 
     fn store(&self) -> &KeyValueStore;
 
+    fn new_store(&self) -> &KeyValueStore;
```

Comment:
From these names it isn't apparent what the difference is between `store()` and `new_store()`.

Reply:
renamed